lib/App/Config.pm

```diff
@@ -0,0 +1,422 @@
+package App::Config;
+
+use strict;
+use warnings;
+use Time::HiRes qw(time);
+
+=head1 NAME
+
+App::Config - An OO configuration module which can be changed and stored into chronicle database.
+
+=head1 VERSION
+
+Version 0.01
+
+=cut
+
+our $VERSION = '0.01';
+
+=head1 SYNOPSIS
+
+    my $app_config = App::Config->new;
+
+=head1 DESCRIPTION
+
+This module parses configuration files and provides interface to access
+configuration information.
+
+=head1 FILE FORMAT
+
+The configuration file is a YAML file. Here is an example:
+
+    system:
+      description: "Various parameters determining core application functionality"
+      isa: section
+      contains:
+        email:
+          description: "Dummy email address"
+          isa: Str
+          default: "dummy@mail.com"
+          global: 1
+        admins:
+          description: "Are we on Production?"
+          isa: ArrayRef
+          default: []
+
+Every attribute is very intuitive. If an item is global, you can change its value and the value will be stored into chronicle database by calling the method C<save_dynamic>.
+
+=head1 SUBROUTINES/METHODS
+
+=cut
+
+use Moose;
+use namespace::autoclean;
+use YAML::XS qw(LoadFile);
+
+use App::Config::Attribute::Section;
+use App::Config::Attribute::Global;
+use Data::Hash::DotNotation;
+
+use Data::Chronicle::Reader;
+use Data::Chronicle::Writer;
+
+=head2 definition_yml
+
+The YAML file that store the configuration
+
+=cut
+
+has definition_yml => (
+    is       => 'ro',
+    isa      => 'Str',
+    required => 1,
+);
+
+=head2 chronicle_reader
+
+The chronicle store that configurations can be fetch from it. It should be an instance of L<Data::Chronicle::Reader>.
+
+=cut
+
+has chronicle_reader => (
+    is       => 'ro',
+    isa      => 'Data::Chronicle::Reader',
```

Comment:
for now, this is probably ok. But I'd suggest to describe the interface the chronicle object has to provide and accept any object here. That would make you independent of our Chronicle module. You can mention in the description that `Data::Chronicle` implements exactly that interface. But the user is free to implement any storage backend he wants.